File: src/commands/play_command.cpp
#include "commands/play_command.hpp"

#include <plog/Log.h>

#include <iostream>

#include "utils.hpp"

PlayCommand::PlayCommand(const std::shared_ptr<StationsDao> stationsDao, 
                         const std::shared_ptr<SettingsDao> settingsDao,
                         const std::shared_ptr<MediaPlayer> mediaPlayer)
    : Command("play", "Play a station, given by id", stationsDao, settingsDao, mediaPlayer), m_waitOnPlay(true) { }

void PlayCommand::execute(const std::vector<std::string>& args) {
    m_cli.parse(args);

    if (m_cli.hasOption('h')) {
        LOG(plog::info) << m_cli.usage();
        return;
    }

    if (m_cli.getResidualValues().empty()) {
        LOG(plog::error) << "No id given. See 'cora play --help' for more information.";
        return;
    }

    // First entry is the command
    const std::string idAndUrl = m_cli.getResidualValues()[1];    
    const std::vector<std::string> values = split(idAndUrl, ':');
    std::shared_ptr<Station> station = getStation(values);

    if (station != nullptr) {
        LOG(plog::info) << "playing " << station->getName();
        std::string url = getUrl(station, values);
        play(url);
    } else {
        LOG(plog::warning) << "No station found for id:" << values[0];
    }
}

std::shared_ptr<Station> PlayCommand::getStation(const std::vector<std::string>& values) const {
    LOG(plog::debug) << "checking: " << values[0];
    m_stationsDao->open(m_cli.getValue('f', getDefaultFile()));
    m_settingsDao->open(m_cli.getValue('f', getDefaultFile()));
    std::shared_ptr<Station> station;

    if (values[0].compare("random") == 0) {
        station = m_stationsDao->getRandom();
    } else if (values[0].compare("last") == 0) {
        std::string lastPlayed = m_settingsDao->get(Settings::LAST_PLAYED);
        if (lastPlayed.empty()) {
            LOG(plog::info) << "There is no last played station. Selecting random.";
            station = m_stationsDao->getRandom();
        } else {
            const long id = std::stol(lastPlayed);
            station = m_stationsDao->findById(id);
        }
    } else {
        const long id = std::stol(values[0]);
        station = m_stationsDao->findById(id);
    }

    m_settingsDao->save(Settings::LAST_PLAYED, std::to_string(station->getId()));
    m_stationsDao->close();
    return station;
}

std::string PlayCommand::getUrl(std::shared_ptr<Station> station, const std::vector<std::string>& values) const {
    std::string url;
    if (values.size() > 1) {
        const size_t index = std::stol(values[1]);

        if (station->getUrls().size() < index) {
            LOG(plog::warning) << "Only " << station->getUrls().size() << " found. Index " << index << " is invalid";
            LOG(plog::info) << "Using default index 0";
            url = station->getUrls()[0];
        } else {
            url = station->getUrls()[index];
        }
    } else {
        url = station->getUrls()[0];
    }

    return url;
}

void PlayCommand::play(const std::string& url) const {
    LOG(plog::debug) << "url: " << url;
    m_mediaPlayer->setUrl(url);
    m_mediaPlayer->play();

    if (m_waitOnPlay) { // For easy testing this can be deactivated
        LOG(plog::info) << "Press enter to stop playing";
        std::cin.get();
    }

    m_mediaPlayer->stop();
}

void PlayCommand::waitOnPlay(const bool waitOnPlay) {
    m_waitOnPlay = waitOnPlay;
}

File: cora/play_command.py

import argparse
import logging

import command

__LAST_PLAYED__ = "last.played"

logger = logging.getLogger(__name__)


class PlayCommand(command.Command):

    def __init__(self, stations, settings, player, subparsers):
        super().__init__(stations, settings, player)
        parser = subparsers.add_parser('play', help='Play a station given by ID or list of keywords')
        parser.add_argument('keywords', nargs=argparse.REMAINDER, help='An ID or a list of keywords')
        parser.set_defaults(func=self.execute)

    def execute(self, args):
        logger.debug("keywords: " + str(args.keywords))

        try:
            id = int(args.keywords[0])
            station = self.stations.find_by_id(id)
        except ValueError:
            logger.debug(args.keywords[0] + " is not an id")
            if args.keywords[0] == 'last':
                id = self.settings.get(__LAST_PLAYED__)
                logger.debug("last played: " + str(id))
                station = self.stations.find_by_id(id)
            else:
                keywords = '%' + '%'.join(args.keywords) + '%'
                logger.debug("keywords: " + keywords)
                station = self.stations.find_by_keywords(keywords)

        if station is None:
            logger.info("No station found for " + str(args.keywords))
        else:
            logger.info("playing " + station.name.strip())
            try:
                self.player.set_url(station.urls[0])
                self.settings.save(__LAST_PLAYED__, station.id)
                self.player.play()
                playing = True
                while playing:
                    logger.info("Press any key to stop playing")
                    key = input()
                    playing = False
            except KeyboardInterrupt:
                # Handle interrupt nicely
                pass
            self.player.stop()
